File: 309-best-time-to-buy-and-sell-stock-with-cooldown/309-best-time-to-buy-and-sell-stock-with-cooldown.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        dp = [-prices[0],0,0]
        
        for i in range(1,len(prices)):
            temp = dp.copy()
            dp = [0,0,0]
            dp[0] = max(temp[0],temp[2]-prices[i])
            dp[1] = temp[0] + prices[i]
            dp[2] = max(temp[2],temp[1])
            print(dp)
        return max(dp)
        
        

File: 309-best-time-to-buy-and-sell-stock-with-cooldown/309-best-time-to-buy-and-sell-stock-with-cooldown.java
class Solution {
    public int maxProfit(int[] prices) {
        int[] dp = new int[3] ;
        dp[0] = -prices[0];
        int[] temp ;
        for (int i = 1 ; i < prices.length; i ++){
            temp = dp.clone();
            dp[0] = Math.max(temp[0],temp[2] - prices[i]);
            dp[1] = Math.max(temp[1],temp[0] + prices[i]) ; 
            dp[2] = Math.max(temp[1],temp[2]);
        }
        return Math.max(dp[1],dp[2]) ;
        
    }
}